util/ioutil.go

```diff
@@ -0,0 +1,42 @@
+// Part of ndt-server-go <https://github.com/m-lab/ndt-server-go>, which
+// is free software under the Apache v2.0 License.
+
+package util
+
+import (
+	"io"
+	"net"
+	"time"
+)
+
+// DeadlineReader is a io.Reader where the read has a deadline. See also
+// <https://groups.google.com/forum/#!topic/golang-nuts/afgEYsoV8j0>
+type DeadlineReader struct {
+	io.Reader
+	conn     net.Conn
+	deadline time.Duration
+}
+
+func NewDeadlineReader(conn net.Conn) DeadlineReader {
+	return DeadlineReader{
+		conn:     conn,
+		deadline: 10.0 * time.Second,
+	}
+}
+
```

Comment:
Or we could (in future) override io.Read

func (rdr *DeadlineReader) Read(p []byte) (n int, err error) {
   // maybe adjust deadline
   // ...
   // Then execute the read.
   return rdr.Reader.Read(p)
}

Then we can just directly call ReadFull(dr, data)